src/lib/4337.js

```diff
@@ -0,0 +1,249 @@
+const erc20Token = new ethers.utils.Interface(tokenContractAbi);
+
+const adjustForDecimals = (amount, decimals) => {
+  return amount * 10 ** decimals;
+};
+
+const transferCallData = (tokenAddress, receiver, amount) =>
+  accountInterface.encodeFunctionData("execute", [
+    tokenAddress,
+    ethers.constants.Zero,
+    erc20Token.encodeFunctionData("transfer", [receiver, amount]),
+  ]);
+
+const userOpToJson = (userop) => {
+  const newUserop = {
+    sender: userop.sender,
+    nonce: userop.nonce.toHexString().replace("0x0", "0x"),
+    initCode: ethers.utils.hexlify(userop.initCode),
+    callData: ethers.utils.hexlify(userop.callData),
+    callGasLimit: userop.callGasLimit.toHexString().replace("0x0", "0x"),
+    verificationGasLimit: userop.verificationGasLimit
+      .toHexString()
+      .replace("0x0", "0x"),
+    preVerificationGas: userop.preVerificationGas
+      .toHexString()
+      .replace("0x0", "0x"),
+    maxFeePerGas: userop.maxFeePerGas.toHexString().replace("0x0", "0x"),
+    maxPriorityFeePerGas: userop.maxPriorityFeePerGas
+      .toHexString()
+      .replace("0x0", "0x"),
+    paymasterAndData: ethers.utils.hexlify(userop.paymasterAndData),
+    signature: ethers.utils.hexlify(userop.signature),
+  };
+
+  return newUserop;
+};
+
+const userOpFromJson = (userop) => {
+  const newUserop = {
+    sender: userop.sender,
+    nonce: ethers.BigNumber.from(userop.nonce),
+    initCode: ethers.utils.arrayify(userop.initCode),
+    callData: ethers.utils.arrayify(userop.callData),
+    callGasLimit: ethers.BigNumber.from(userop.callGasLimit),
+    verificationGasLimit: ethers.BigNumber.from(userop.verificationGasLimit),
+    preVerificationGas: ethers.BigNumber.from(userop.preVerificationGas),
+    maxFeePerGas: ethers.BigNumber.from(userop.maxFeePerGas),
+    maxPriorityFeePerGas: ethers.BigNumber.from(userop.maxPriorityFeePerGas),
+    paymasterAndData: ethers.utils.arrayify(userop.paymasterAndData),
+    signature: ethers.utils.arrayify(userop.signature),
+  };
+
+  return newUserop;
+};
+
+const senderAccountExists = async (indexer, sender) => {
+  const url = `${indexer.url}/accounts/${sender}/exists`;
+
+  const resp = await fetch(url);
+
+  return resp.status === 200;
+};
+
+const generateUserOp = (
+  signerAddress,
+  sender,
+  senderAccountExists = false,
+  accountFactoryAddress,
+  callData
+) => {
+  const userop = {
+    sender,
+    nonce: ethers.BigNumber.from("0"),
+    initCode: ethers.utils.arrayify("0x"),
+    callData: ethers.utils.arrayify("0x"),
+    callGasLimit: ethers.BigNumber.from("0"),
+    verificationGasLimit: ethers.BigNumber.from("1500000"),
+    preVerificationGas: ethers.BigNumber.from("21000"),
+    maxFeePerGas: ethers.BigNumber.from("0"),
+    maxPriorityFeePerGas: ethers.BigNumber.from("1000000000"),
+    paymasterAndData: ethers.utils.arrayify("0x"),
+    signature: ethers.utils.arrayify("0x"),
+  };
+
+  // initCode
+  if (!senderAccountExists) {
```

Comment:
why do we care if the sender is an EOA?

Reply:
This to check if an account is deployed. If it isn't, include the deployment code in the userop.

Standard 4337 stuff.

Reply:
Normally you would use the nonce, but since we have started using a more extravagant nonce, it's easier to just check if an account is deployed.